File: 09_Recursion/M039. Combination Sum.java
// Combination Sum
// https://leetcode.com/problems/combination-sum/

import java.util.*;

class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);  // sort for pruning
        List<List<Integer>> result = new ArrayList<>();
        backtrack(result, new ArrayList<>(), candidates, target, 0);
        return result;
    }
    
    private void backtrack(List<List<Integer>> result, List<Integer> current, int[] candidates, int target, int start) {
        if (target == 0) {
            result.add(new ArrayList<>(current)); // only copy here
            return;
        }
            
        for (int i = start; i < candidates.length; i++) {
            int val = candidates[i];
            if (val > target) break;  
                
            current.add(val);
            backtrack(result, current, candidates, target - val, i);
            current.remove(current.size() - 1);
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] candidates = { 2, 3, 6, 7 };
        int target = 7;
        List<List<Integer>> result = solution.combinationSum(
            candidates,
            target
        );
        System.out.println(result);
    }
}

File: JavaScript/09_Recursion/M039. Combination Sum.js
// 39. Combination Sum
// https://leetcode.com/problems/combination-sum/

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
function combinationSum(candidates, target) {
    candidates.sort((a, b) => a - b); // sort for pruning
    const result = [];
    backtrack(result, [], candidates, target, 0);
    return result;
}

function backtrack(result, current, candidates, target, start) {
    if (target === 0) {
        result.push([...current]); // only copy here
        return;
    }

    for (let i = start; i < candidates.length; i++) {
        const val = candidates[i];
        if (val > target) break;

        current.push(val);
        backtrack(result, current, candidates, target - val, i);
        current.pop();
    }
}

const candidates = [2, 3, 6, 7];
const target = 7;
console.log(combinationSum(candidates, target)); // Output: [[2, 2, 3], [7]]
